### PR TITLE
create_region_changes

### PR DESCRIPTION
**Pull request by soumya-pn**
added region_attributes to fam_create_region().

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -190,7 +190,7 @@ add_subdirectory(test/coverage)
 add_subdirectory(test/microbench)
 add_subdirectory(test/apps)
 add_subdirectory(test/permission-test)
-add_subdirectory(examples/api)
+#add_subdirectory(examples/api)
```
Comment (mashoodk) [inline CMakeLists.txt:193]: You can add this back and resolve issues in test cases.

File: include/fam/fam.h [hunk 2]
```diff
@@ -101,11 +101,25 @@ typedef long long int128_t;
 #endif // __int128t
 #endif // int128_t
 
+typedef enum {
+    INTERLEAVE_DEFAULT,
```
Comment (faizan-barmawer) [inline include/fam/fam.h:105]: we can set the default enum value as zero for all the enum parameters.
INTERLEAVE_DEFAULT=0,

File: include/fam/fam.h [hunk 3]
```diff
@@ -114,6 +128,11 @@ typedef enum {
     RAID5
 } Fam_Redundancy_Level;
 
+typedef struct {
```
Comment (faizan-barmawer) [inline include/fam/fam.h:131]: Might need to properly format the code.
Use clang-format tool to follow the coding convention and make appropriate changes.

File: include/fam/fam.h [hunk 4]
```diff
@@ -226,6 +245,9 @@ class Fam_Region_Descriptor {
     void set_size(uint64_t itemSize);
     void set_perm(mode_t perm);
     void set_name(char *name);
+    void set_redundancyLevel(Fam_Redundancy_Level redundancyLevel);
```
Comment (faizan-barmawer) [inline include/fam/fam.h:248]: Need to think whether we should have the "set_*" methods in the descriptor class. 
We don't want the application to use these methods to set any of the descriptor properties. These methods need to be used only by the implementation(internally).

File: src/fam-api/fam.cpp [hunk 5]
```diff
@@ -3861,9 +3861,26 @@ Fam_Descriptor *fam::fam_lookup(const char *itemName, const char *regionName) {
  */
 Fam_Region_Descriptor *
 fam::fam_create_region(const char *name, uint64_t size, mode_t permissions,
-                       Fam_Redundancy_Level redundancyLevel, ...) {
+                       Fam_Region_Attributes* regionAttributes) {
     TRY_CATCH_BEGIN
-    return pimpl_->fam_create_region(name, size, permissions, redundancyLevel);
+    if(regionAttributes == NULL)
```
Comment (faizan-barmawer) [inline src/fam-api/fam.cpp:3866]: This "regionAttributes" parameter is passed by the application. 
Implementation should only read this parameter. 
If applicaiton call create_region api with "NULL", just use default parameters. 
We need to save these values in the region descriptor, by looking at the input regionAttribute. Either have separate copy of the regionAttribute structure or keep it as separate data items in the region descriptor.

**Review by faizan-barmawer — CHANGES_REQUESTED** [review]
Hi Soumya,

The following changes required:
  - processing of the input region attribute parameter
  - use clang-format and make appropriate code format changes.
 
Regards,
Faizan

File: include/fam/fam.h (hunk 3 above)
Comment (soumya-pn) [reply include/fam/fam.h:131]: addressed

File: include/fam/fam.h (hunk 4 above)
Comment (soumya-pn) [reply include/fam/fam.h:248]: As discussed raised an issue to track this

File: src/fam-api/fam.cpp (hunk 5 above)
Comment (soumya-pn) [reply src/fam-api/fam.cpp]: resolved

File: include/fam/fam.h [hunk 6]
```diff
@@ -101,11 +101,17 @@ typedef long long int128_t;
 #endif // __int128t
 #endif // int128_t
 
+typedef enum { INTERLEAVE_DEFAULT = 0, ENABLE, DISABLE } Fam_Interleave_Enable;
```
Comment (mashoodk) [inline include/fam/fam.h:104]: Add comments - refer to comments in Fam_Redundancy_Level

File: test/common/fam_test_config.h.in [hunk 7]
```diff
@@ -38,9 +38,12 @@
 #include <string.h>
 #include <fam/fam.h>
 
+#define RAID1 NULL
+#define RAID5 NULL
+#define NONE NULL
```
Comment (mashoodk) [inline test/common/fam_test_config.h.in:43]: @faizan-barmawer What do we do in test cases. Should we change all test cases to use NULL in place of NONE ?

**Review by mashoodk — CHANGES_REQUESTED** [review]

File: test/apps/spmv/csr_perload.cpp [hunk 8]
```diff
@@ -28,12 +28,11 @@
  *
  */
 #include "common/fam_options.h"
+#include "common/fam_test_config.h"
```
Comment (mashoodk) [inline test/apps/spmv/csr_perload.cpp:31]: Can you change create_region calls in this file (apps) in place of adding common/fam_test_config.h

**Review by mashoodk — CHANGES_REQUESTED** [review]

**Review by mashoodk — APPROVED** [review]

**Review by faizan-barmawer — APPROVED** [review]

**Pull request merged**